### ISSUE TITLE
[breaking change] removing `dart pub uploader`

### ISSUE THREAD

**Issue by jonasfj** [dart-lang/sdk#48526]
Starting Dart 2.17 the `dart pub uploader add|remove` commands will be removed from the Dart SDK.

```console
$ dart pub uploader --help
Manage uploaders for a package on pub.dartlang.org.

Usage: dart pub uploader [options] {add/remove} <email>
-h, --help               Print this usage information.
    --package            The package whose uploaders will be modified.
                         (defaults to the current package)
-C, --directory=<dir>    Run this in the directory<dir>.

Run "dart help" to see global options.
See https://dart.dev/tools/pub/cmd/pub-uploader for detailed documentation.
```

## Motivation

Package administrative functionality is exposed as part of the web-interface on pub.dev:
`https://pub.dev/packages/<package>/admin`

The package admin page on pub.dev has long supported adding/remove uploaders, keeping the functionality in the client duplicates functionality. Putting all administrative operations into the web-interface makes it easier to introduce and change features outside Dart SDK releases, and increases discoverability of new administrative features introduced.

Most package authors are also better served by registering a [verified publisher](https://dart.dev/tools/pub/verified-publishers) and managing access to packages through said publisher, instead of managing uploaders on a per-package basis.

## Impact

Users running `dart pub uploader` will get a message telling them to use `https://pub.dev/packages/<package>/admin` instead.

## Migration

Use the web-interface at `https://pub.dev/packages/<package>/admin` instead.
This requires sign-in with Google, the same Google account as is used with `dart pub uploader` and `dart pub publish`.

**Comment by JaysonBH** [comment]
Wouldn't this break automation?
Would it now be recommended to use a manual flow instead?
Is there more info on the reason for this decision?
Is it solving an existing problem?

**Comment by jonasfj** [comment]
> Wouldn't this break automation?

If you are using `dart pub uploader` to automatically add/remove uploaders from a package then yes.
But this CLI command is not well suited for automation any ways. It stores a Google oauth2 `refresh_token` in `$XDG_CONFIG_HOME/...` and if the token doesn't work then it initiates an oauth-login flow by printing a URL and listing for the callback on localhost. Hence:
 * `dart pub uploader` assumes a human is sitting at the computer.
 * `dart pub uploader` stores credentials in a config file that most don't know how to find or move.
 * `dart pub uploader add` only sends an invite to become uploader, so human interaction / confirmation is required by the invited uploader.
 * This command cannot use service accounts for authentication.



> Would it now be recommended to use a manual flow instead?

Yes, using `https://pub.dev/packages/<package>/admin` is the recommended mechanism for inviting new uploaders.

But if you're adding/removing uploaders for a large collection of packages managed by the same set of users, you really should be using a [verified publisher](https://dart.dev/tools/pub/verified-publishers). When using verified publishers multiple packages can be owned by a single publisher, and all members of the publisher can manage the packages. For managing members of a publisher you have to use the web-interface.


> Is there more info on the reason for this decision?

<details>
<summary>TL;DR: Decoupling the interface for managing packages from the `dart pub` CLI client grants us more freedom to innovate and allows us to iterate faster.</summary>

As a general principle it makes sense to decouple administration of packages from the CLI client. It allows us to increase velocity and ship features out-of-band. For example, we've shipped features like "[retraction of package version](https://dart.dev/tools/pub/publishing#retract)", "[verified publishers](https://dart.dev/tools/pub/verified-publishers)" and "[flagging discontinued](https://dart.dev/tools/pub/publishing#discontinue)" without support in the `dart pub` CLI client; this made it easier to ship and faster (this is especially nice for small features like "flagging discontinued"). 

Moreover, shipping features in the `dart pub` CLI client means that users have to upgrade to a new version of the Dart SDK in order to benefit from those features. For many features such as managing uploaders, flagging a package as discontinued, etc. There is no need changes in the `dart pub` CLI client.

In general, it's harder to ship administrative interface in the CLI client, because we can't quickly fix minor bugs or annoyances after we've shipped. With administrative interface through the web-interface the round-trip for improving a text message that explains something is much shorter.

The final reason is that there might be other pub servers out there. We've documented the specification for these servers in [doc/repository-spec-v2.md](https://github.com/dart-lang/pub/blob/master/doc/repository-spec-v2.md). If we put administrative interface in the `dart pub` CLI client, users might expect such administrative interface to also work for custom pub servers, and _some_ developers of custom pub servers might want the [repository specification](https://github.com/dart-lang/pub/blob/master/doc/repository-spec-v2.md) expanded to cover administrative actions like `dart pub uploader add`. Again this would be non-trivial work, it would again make it harder to make changes in how packages/permissions are managed, because said logic would be:
 * codified in a specification,
 * shipped in installed clients (`dart pub` CLI client), and,
 * implemented by custom package servers.

Putting permission/package management features in the `dart cli` client would not only make it harder to change in the future (and harder to introduce new concepts like user roles for members of a _verified publisher_). But it would also force the permission model we design onto authors of custom package repositories. For example: [cloudsmith](https://help.cloudsmith.io/docs/dart-repository) and [Jfrog Artifactory](https://jfrog.com/blog/how-to-use-pub-repositories-in-artifactory/) (both of which have Dart support) have different ways to manage permissions and such custom repositories would typically allow users to integrate with SSO (maybe even support managing permissions through an existing LDAP system, or similar).

The point is: to the extent possible it is probably preferable to keep the API exposed through `dart pub` client as small as possible. If people really need to automate management against pub.dev, we should expose web APIs specific to pub.dev and allow authentication using service accounts.

</details>

**Comment by itsjustkevin** [comment]
@jonasfj could you confirm that this breaking change is already in dart 2.17?

**Comment by jonasfj** [comment]
@itsjustkevin yes, this landed in 2.17 -- I guess we can close this issue.
